src/mlpack/core/data/split_data.hpp

```diff
@@ -52,56 +54,27 @@ namespace data {
  * @param shuffleData If true, the sample order is shuffled; otherwise, each
  *     sample is visited in linear order. (Default true.)
  */
-template<typename T, typename U>
+template<typename T, typename LabelsType,
+         typename = std::enable_if_t<arma::is_arma_type<LabelsType>::value> >
 void StratifiedSplit(const arma::Mat<T>& input,
-                     const arma::Row<U>& inputLabel,
+                     const LabelsType& inputLabel,
                      arma::Mat<T>& trainData,
                      arma::Mat<T>& testData,
-                     arma::Row<U>& trainLabel,
-                     arma::Row<U>& testLabel,
+                     LabelsType& trainLabel,
+                     LabelsType& testLabel,
                      const double testRatio,
                      const bool shuffleData = true)
 {
-  /**
-   * Basic idea:
-   * Let us say we have to stratify a dataset based on labels:
-   * 0 0 0 0 0 (5 0s)
-   * 1 1 1 1 1 1 1 1 1 1 1 (11 1s)
-   *
-   * Let our test ratio be 0.2.
-   * Then, the number of 0 labels in our test set = floor(5 * 0.2) = 1.
-   * The number of 1 labels in our test set = floor(11 * 0.2) = 2.
-   *
-   * In our first pass over the dataset,
-   * We visit each label and keep count of each label in our 'labelCounts' uvec.
-   *
-   * We then take a second pass over the dataset.
-   * We now maintain an additional uvec 'testLabelCounts' to hold the label
-   * counts of our test set.
-   *
-   * In this pass, when we encounter a label we check the 'testLabelCounts' uvec
-   * for the count of this label in the test set.
-   * If this count is less than the required number of labels in the test set,
-   * we add the data to the test set and increment the label count in the uvec.
-   * If this count is equal to or more than the required count in the test set,
-   * we add this data to the train set.
-   *
-   * Based on the above steps, we get the following labels in the split set:
-   * Train set (4 0s, 9 1s)
-   * 0 0 0 0
-   * 1 1 1 1 1 1 1 1 1
-   *
-   * Test set (1 0s, 2 1s)
-   * 0
-   * 1 1
-   */
+  if (!arma::is_Row<LabelsType>::value)
+    throw std::runtime_error("data::Split(): when stratified sampling is done,"
+        "labels must have type `arma::Row<>`!");
   size_t trainIdx = 0;
   size_t testIdx = 0;
   size_t trainSize = 0;
   size_t testSize = 0;
   arma::uvec labelCounts;
   arma::uvec testLabelCounts;
-  U maxLabel = inputLabel.max();
+  auto maxLabel = inputLabel.max();
```

Comment:
Do you mind to use `typename LabelsType::elem_type` here, for more details see the armadillo section about "Can I use the C++11 auto keyword with Armadillo objects and/or expressions?".

Reply:
Thanks for pointing this out!.